WalletWasabi.Gui/Rpc/JsonRpcRequestHandler.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using WalletWasabi.JsonConverters;
+
+namespace WalletWasabi.Gui.Rpc
+{
+    ///<summary>
+    /// This class coordinates all the major steps in processing the RPC call.
+    /// It parses the json request, parses the parameters, invokes the service
+    /// methods and handles the errors.
+    ///</summary>
+    public class JsonRpcRequestHandler<TService>
+    {
+        private static readonly JsonSerializerSettings DefaultSettings = new JsonSerializerSettings
+        {
+            NullValueHandling = NullValueHandling.Ignore,
+            ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
+            Converters = new JsonConverter[] {
+                new Uint256JsonConverter(),
+                new OutPointJsonConverter(),
+                new BitcoinAddressJsonConverter()
+            }
+        };
+
+        private static readonly JsonSerializer DefaultSerializer = JsonSerializer.Create(DefaultSettings);
+
+        private TService Service { get; }
+        private JsonRpcServiceMetadataProvider MetadataProvider { get; }
+
+        public JsonRpcRequestHandler(TService service)
+        {
+            Service = service;
+            MetadataProvider = new JsonRpcServiceMetadataProvider(typeof(TService));
+        }
+
+        /// <summary>
+        /// Parses the request and dispatches it to the correct service's method.
+        /// </summary>
+        /// <param name="body">The raw rpc request.</param>
+        /// <param name="cancellationToken">The cancellation token that will be past to the service handler in case it expects/accepts one.</param>
+        /// <returns>The response that, after serialization, is returned as response.</returns>
+        public async Task<string> HandleAsync(string body, CancellationToken cancellationToken)
```

Comment:
I would separate the parser and the dispatch. The reason is to write some tests for the parser - it seems to be complex.

Reply:
The parsing logic is trivial and it is in the `JsonRpcRequest.TryParse` method. Here we don't deal with parsing logic, what we do is simply return an jsonrpc error in case the request is incorrect.